File: python/src/application/user_profile_service.py
import random

from src.application import id_generator
from src.application.application_exception import ApplicationException
from src.application.lock import lock_keys
from src.application.lock.lock_service import LockService
from src.domain.sms_gateway import SmsGateway
from src.domain.user import User
from src.domain.user_repository import UserRepository
from src.domain.user_unique_checker import UserUniqueChecker
from src.utility import value_utility


class UserProfileService:
    """用户资料Service"""

    def __init__(self, user_repository: UserRepository, user_unique_checker: UserUniqueChecker,
                 sms_gateway: SmsGateway, lock_service: LockService) -> None:
        self.user_repository = user_repository
        self.user_unique_checker = user_unique_checker
        self.sms_gateway = sms_gateway
        self.lock_service = lock_service
        self.mobile_code: dict[str, str] = {}

    async def register(self, username: str, password: str, nickname: str) -> str:
        """注册，仅演示使用，未防止恶意注册"""
        async with self.lock_service.lock(lock_keys.USER):
            user = await User.register_user(
                id=id_generator.generate_id(),
                username=username,
                password=password,
                nickname=nickname,
                user_unique_checker=self.user_unique_checker)
            await self.user_repository.insert(user)
            return user.id

    async def modify_password(self, user_id: str, old_password: str, new_password: str) -> None:
        """修改密码"""
        user = await self.user_repository.select_by_id_req(user_id)
        user.modify_password(old_password, new_password)
        await self.user_repository.update(user)

    async def modify_nickname(self, user_id: str, nickname: str) -> None:
        """修改昵称"""
        async with self.lock_service.lock(lock_keys.USER):
            user = await self.user_repository.select_by_id_req(user_id)
            await user.modify_nickname(nickname, self.user_unique_checker)
            await self.user_repository.update(user)

    async def send_mobile_code(self, user_id: str, mobile: str) -> None:
        """发送手机验证码"""
        if value_utility.is_blank(mobile):
            raise ApplicationException("手机号不能为空")
        user = await self.user_repository.select_by_id_req(user_id)
        code = str(random.randint(100000, 999999))
        self.mobile_code[user.id + "_" + mobile] = code
        await self.sms_gateway.send_code(mobile, code)

        # import asyncio
        # asyncio.create_task(self.sms_gateway.send_code(mobile, code))

    async def bind_mobile(self, user_id: str, mobile: str, code: str) -> None:
        """绑定手机"""
        if value_utility.is_blank(code):
            raise ApplicationException("验证码不能为空")
        user = await self.user_repository.select_by_id_req(user_id)
        key = user.id + "_" + mobile
        if self.mobile_code.get(key, "") != code:
            raise ApplicationException("手机验证码错误")
        await user.modify_mobile(mobile, self.user_unique_checker)
        await self.user_repository.update(user)
        self.mobile_code.pop(key, None)

File: dotnet/src/application/UserProfileService.cs
using System;
using System.Collections.Concurrent;
using BackendExample.Domain;
using BackendExample.Utility;

namespace BackendExample.Application
{
    /// <summary>
    /// 用户资料 Service
    /// </summary>
    public class UserProfileService
    {
        private readonly UserRepository userRepository;
        private readonly UserUniqueSpecification userUniqueSpecification;
        private readonly SmsGateway smsGateway;
        private readonly LockService lockService;

        private ConcurrentDictionary<string, string> mobileCodeDict = new ConcurrentDictionary<string, string>();

        public UserProfileService(
            UserRepository userRepository,
            UserUniqueSpecification userUniqueSpecification,
            SmsGateway smsGateway,
            LockService lockService
            )
        {
            this.userRepository = userRepository;
            this.userUniqueSpecification = userUniqueSpecification;
            this.smsGateway = smsGateway;
            this.lockService = lockService;
        }

        /// <summary>
        /// 注册
        /// </summary>
        public async Task<string> Register(string username, string password, string nickname)
        {
            return await this.lockService.GetWithLockAsync(LockKeys.USER, async () =>
            {
                User user = await User.RegisterUser(IdGenerator.GenerateId(), username, password, nickname,
                    this.userUniqueSpecification);
                await this.userRepository.Insert(user);
                return user.Id;
            });
        }

        /// <summary>
        /// 修改密码
        /// </summary>
        public async Task ModifyPassword(string userId, string oldPassword, string newPassword)
        {
            User user = await this.userRepository.SelectByIdReq(userId);
            user.ModifyPassword(oldPassword, newPassword);
            await this.userRepository.Update(user);
        }

        /// <summary>
        /// 修改昵称
        /// </summary>
        public async Task ModifyNickname(string userId, string nickname)
        {
            await this.lockService.RunWithLockAsync(LockKeys.USER, async () =>
            {
                User user = await this.userRepository.SelectByIdReq(userId);
                await user.ModifyNickname(nickname, this.userUniqueSpecification);
                await this.userRepository.Update(user);
            });
        }

        /// <summary>
        /// 发送手机验证码
        /// </summary>
        public async Task SendMobileCode(string userId, string mobile)
        {
            if (ValueUtility.IsBlank(mobile))
                throw new ApplicationException("手机号不能为空");

            User user = await this.userRepository.SelectByIdReq(userId);
            string code = new Random().Next(100000, 1000000).ToString();
            this.mobileCodeDict.AddOrUpdate(user.Id + "_" + mobile, code, (k, v) => code);
            await this.smsGateway.SendCode(mobile, code);
        }

        /// <summary>
        /// 绑定手机
        /// </summary>
        public async Task BindMobile(string userId, string mobile, string code)
        {
            if (ValueUtility.IsBlank(code))
                throw new ApplicationException("验证码不能为空");

            User user = await this.userRepository.SelectByIdReq(userId);
            string key = user.Id + "_" + mobile;
            if (!this.mobileCodeDict.TryGetValue(key, out string? sendCode) || sendCode != code)
                throw new ApplicationException("验证码错误");

            await user.ModifyMobile(mobile, this.userUniqueSpecification);
            await this.userRepository.Update(user);
            this.mobileCodeDict.TryRemove(key, out _);
        }
    }
}
